### PR TITLE
OPRUN-3458,OPRUN-3459: add policy enforcement for bundle metadata format based on OCP versions

### PR DESCRIPTION
**Pull request by grokspawn**
This update augments the fbc-validation task to enforce bundle metadata formats for incoming fragments based on the destination catalog's OCP release.

For OCP releases before 4.17, this will ensure that all bundles use only the `olm.bundle.object` object.  
For OCP releases at/after 4.17, this will ensure that all bundles use only the `olm.csv.metadata` object.

### CODE REVIEW

File: task/fbc-validation/0.1/fbc-validation.yaml
Comment (joelanford) [inline task/fbc-validation/0.1/fbc-validation.yaml:77]: I'd suggest colocating these configurations with the check that uses them.  But I'd also defer to the maintainers if they have different opinions.

File: task/fbc-validation/0.1/fbc-validation.yaml
Comment (joelanford) [inline task/fbc-validation/0.1/fbc-validation.yaml:185]:
```suggestion
            echo "!FAILURE! - olm.bundle.object bundle properties are not permitted in a FBC fragment for OCP version ${OCP_VER_MAJOR}.${OCP_VER_MINOR}. Fragments must move to olm.csv.metadata bundle metadata."
```

Should we also include something like this:
```
Use opm >= 1.46.0 with `--migrate-level=bundle-object-to-csv-metadata`
```

File: task/fbc-validation/0.1/fbc-validation.yaml
Comment (joelanford) [inline task/fbc-validation/0.1/fbc-validation.yaml:194]:
```suggestion
            echo "!FAILURE! - olm.csv.metadata bundle properties are not permitted in a FBC fragment for OCP version ${OCP_VER_MAJOR}.${OCP_VER_MINOR}. Fragments must only use olm.bundle.object bundle metadata."
```

And then similar suggestion as above for telling folks how to fix it?

File: task/fbc-validation/0.1/fbc-validation.yaml
Comment (grokspawn) [reply task/fbc-validation/0.1/fbc-validation.yaml:185]: I didn't want to too-closely couple the tools, so I stopped short of providing guidance on how to resolve the issue.  I presumed that the first thing they would do is do a `opm $command --help` for the opm roles in production of their fragment.

Comment (joelanford) [reply task/fbc-validation/0.1/fbc-validation.yaml]: But I think people will need to be told:
1. It is (almost always) `opm` that controls which of these properties is written.
2. _If_ you are using `opm`, we recommend 1.46.0 with `--migration-level` pinning.

**Review by arewm — APPROVED** [review]

File: task/fbc-validation/0.1/README.md
Comment (arewm) [inline task/fbc-validation/0.1/README.md:4]: README modifications will be overwritten once someone auto-generates the files using the script at https://github.com/konflux-ci/build-definitions/blob/main/hack/generate-readme.sh

Comment (grokspawn) [reply task/fbc-validation/0.1/README.md:4]: Ah.  Is there a good place to capture this info, or would I be better off omitting it entirely here?

**Comment by grokspawn** [comment]
/hold 
want to ensure this doesn't merge until I get a chance to resolve Joe's comments

File: task/fbc-validation/0.1/README.md
Comment (arewm) [reply task/fbc-validation/0.1/README.md:4]: We can create another file in the task directory. If we want it in the README then we can modify the `generate-readme.sh` to generate the README from multiple files.

The description line that I highlighted _should_ be auto-generated from the description in the task (once you update that to the desired text). The remaining content would need to be pulled in from elsewhere.

Comment (grokspawn) [reply task/fbc-validation/0.1/README.md:4]: Got it.  Thanks, I was unfamiliar with the generation.  I'll see if I can re-flow into the task description, though I'll have to substantially change the formatting. 😄

Comment (grokspawn) [reply task/fbc-validation/0.1/README.md:4]: @arewm 
Looking back on this, and I was sure I'd seen other, manual updates to the README.  So I went back and found https://github.com/konflux-ci/build-definitions/pull/541 for e.g. 

Do you prefer that I attempt to integrate all the contributions to the task description?  It seems that the intent isn't to re-generate the README.md.

**Review by jsztuka — APPROVED** [review]
lgtm, +1 for discussion on failure format.

File: task/fbc-validation/0.1/fbc-validation.yaml
Comment (dirgim) [inline task/fbc-validation/0.1/fbc-validation.yaml:188]: How expensive is the opm render call? 
We can maybe consider rendering once and saving to a JSON file so we can run these kinds of checks on that file instead.

File: task/fbc-validation/0.1/README.md
Comment (arewm) [reply task/fbc-validation/0.1/README.md:4]: We discussed on the architecture call today. I think that we should leave the README as auto-generated and then create additional USAGE.md and/or TROUBLESHOOTING.md files as needed for those use cases. 

Long term, we can look to merge the other files into README.md with the script for completeness.

File: task/fbc-validation/0.1/fbc-validation.yaml
Comment (grokspawn) [reply task/fbc-validation/0.1/fbc-validation.yaml:188]: They're linear based on the number of provided channels/bundles in the fragment.  I'd like to render once and re-examine, but was not sure where to drop the interim render output.  I'll look for existing convention somewhere, unless you can direct me.

File: task/fbc-validation/0.1/fbc-validation.yaml
Comment (dirgim) [inline task/fbc-validation/0.1/fbc-validation.yaml:185]: This check was dropped as part of [KFLUXBUGS-1554](https://issues.redhat.com/browse/KFLUXBUGS-1554), so now the only check is the one above which checks if there's at least one package defined in the rendered catalog.

Comment (grokspawn) [reply task/fbc-validation/0.1/fbc-validation.yaml:185]: My bad.  When I was resolving rebase conflicts this looked like an additional check, but I'll drop the older check in favor of this.

File: task/fbc-validation/0.1/fbc-validation.yaml
Comment (grokspawn) [reply task/fbc-validation/0.1/fbc-validation.yaml]: I believe I resolved these comments with details in the new TROUBLESHOOTING.md file. @joelanford PTAL, and let me know if these are not resolved.

File: task/fbc-validation/0.1/fbc-validation.yaml
Comment (grokspawn) [reply task/fbc-validation/0.1/fbc-validation.yaml]: Since I'd added info to the USAGE.md doc about the original check, I replaced with info about the new check instead: https://github.com/konflux-ci/build-definitions/pull/1327/files#diff-a87221e9ea400bc7daf673722fdfb1fc2a538c2d066e61e284e52a0385129e95R16-R17

File: task/fbc-validation/0.1/USAGE.md
Comment (grokspawn) [inline task/fbc-validation/0.1/USAGE.md:17]:
```suggestion
To validate that at least one package is included in the fragment, the test renders the FBC using `opm` and uses `jq` to count instances of `olm.package` and fails if there are none.
```

nit... *sigh* 

I'll wait a bit to fix this, because I don't feel it's contentious but I had to force-push a bunch to resolve current responses and to keep this PR up-to-date with main.

File: task/fbc-validation/0.1/README.md
Comment (arewm) [inline task/fbc-validation/0.1/README.md:6]: Can we auto-generate this? Something like
```
if [ -f $(dirname $RESOURCE)/USAGE.md ]; then
  echo
  echo -n "For further information on how to use the task, see the USAGE.md file."
fi
if [ -f $(dirname $RESOURCE)/TROUBLESHOOTING.md ]; then
  echo
  echo -n "For troubleshooting assistance, see the TROUBLESHOOTING.md file."
fi
if [ -f $(dirname $RESOURCE)/USAGE.md ] | [ -f $(dirname $RESOURCE)/TROUBLESHOOTING.md ]; then
  echo
fi
```
between these lines: https://github.com/konflux-ci/build-definitions/blob/main/hack/generate-readme.sh#L10-L11 ?

Comment (grokspawn) [reply task/fbc-validation/0.1/README.md:6]: You don't want me to include this information [here](https://github.com/konflux-ci/build-definitions/pull/1327/files#diff-b43bfce7caedfc108e2678a44c111a52a34313d0a2cc255f8a835df82b8610f0R5-R6)?
As part of `task.spec.description`, this information would be populated here already. 

I wasn't sure if folks already had easy access to the markdown files, so I was trying to ensure that the task itself held the necessary references.

Comment (arewm) [reply task/fbc-validation/0.1/README.md:6]: Oh, I see that you put it in the task description itself. I didn't notice that before. I think that we would eventually like to render the build-definitions makdown files in konflux-ci.dev/docs. If we establish the pattern in the markdown doc construction then we wouldn't have to include the repeated information in every task description.

The primary desire for the change was to prevent the README from getting overwritten. That won't happen with your change, so this is now a more minot nitpick.

Comment (grokspawn) [reply task/fbc-validation/0.1/README.md]: I added it to the automation, but stopped short of regenerating the README.md file from it.  There are changes that I lack context to accept/refuse there.  In addition to whitespace changes which only matter to folks viewing the raw content, it also drops the 'Source repository...' and ['Additional links'](https://github.com/konflux-ci/build-definitions/pull/1327/files#diff-b43bfce7caedfc108e2678a44c111a52a34313d0a2cc255f8a835df82b8610f0L27-L30) sections.

**Review by arewm — APPROVED** [review]

**Comment by grokspawn** [comment]
I'm not clear on how to kick the CI.... I see that the CI is not running because 
> User grokspawn is not allowed to trigger CI via pull_request on this repo.

Help?

**Comment by grokspawn** [comment]
🎉 Now it looks like there is an additional required action.  I see an approval.  Is there something I must do?

**Comment by arewm** [comment]
The Checkton check is showing violations in your task changes.

The trusted artifact variant check is likely flagging issues due to your change to the README generator. I know that we have gone back and forth on it, but you can back out that change if you don't want to resolve the differences. That change can be proposed later.

**Comment by grokspawn** [comment]
> The Checkton check is showing violations in your task changes.
> 
> The trusted artifact variant check is likely flagging issues due to your change to the README generator. I know that we have gone back and forth on it, but you can back out that change if you don't want to resolve the differences. That change can be proposed later.

I understand what the tasks are communicating.  The tasks aren't identified by the CI flow as required (and the checkton check didn't exist until 8/30).  

To help folks contribute to this repo, it would be really helpful if 
1. the set of required checks would reflect the merge check minimums; and
2. PRs that introduce checks should also vet impacts to existing code (and IMO create a PR to address, so that other contributors just have to merge/rebase to update their efforts).

**Comment by hongweiliu17** [comment]
The [issue](https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1725429053600649) from #konflux-users might be caused by this commit, see [analysis](https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1725516847492469?thread_ts=1725429053.600649&cid=C04PZ7H0VA8)
The `BASE_IMAGE` is like `registry.redhat.io/openshift4/ose-operator-registry-rhel9@sha256:56e5c57b9d94bcadad6ff0cd47d157d35affc4bfd2c23a2ac948eee1cac3b8a3`, not `registry.redhat.io/openshift4/ose-operator-registry-rhel9:v4.16`, so that the OCP major version and minor version can't be derived correctly.

**Comment by chmeliik** [comment]
>     1. the set of required checks would reflect the merge check minimums; and

I don't think we treat any of the existing checks as optional. PRs don't get merged if the checks are not green.

>     2. PRs that introduce checks should also vet impacts to existing code (and IMO create a PR to address, so that other contributors just have to merge/rebase to update their efforts).

I believe that's the case

**Pull request merged**
